File: 0310-minimum-height-trees/0310-minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1: return [0]

        degree = defaultdict(set)
        result = list()
        que = deque()

        for u, v in edges:
            degree[u].add(v)
            degree[v].add(u)

        for key, value in degree.items():
            if len(value) == 1:
                que.append(key)
                result.append(key)

        while que:
            n = len(que)
            result = list()

            for i in range(n):
                now = que.popleft()
                result.append(now)
                for next_point in degree[now]:
                    degree[next_point].discard(now)
                    if len(degree[next_point]) == 1:
                        que.append(next_point)

        return result

File: 0310-minimum-height-trees/0310-minimum-height-trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> result = new ArrayList<>();
        if (n == 1) {
            result.add(0);
            return result;
        }

        Map<Integer, Set<Integer>> degree = new HashMap<>();
        Queue<Integer> que = new ArrayDeque<>();

        for (int i = 0; i < edges.length; i++) {
            if (!degree.containsKey(edges[i][0])) {
                degree.put(edges[i][0], new HashSet<>());
            }
            if (!degree.containsKey(edges[i][1])) {
                degree.put(edges[i][1], new HashSet<>());
            }
            degree.get(edges[i][0]).add(edges[i][1]);
            degree.get(edges[i][1]).add(edges[i][0]);
        }

        for (int key: degree.keySet()) {
            if (degree.get(key).size() == 1) {
                que.add(key);
                result.add(key);
            }
        }

        while (!que.isEmpty()) {
            int len = que.size();
            result = new ArrayList<>();

            for (int i = 0; i < len; i++) {
                int now = que.poll();
                result.add(now);
                for (int next: degree.get(now)) {
                    degree.get(next).remove(now);
                    if (degree.get(next).size() == 1) {
                        que.add(next);
                    }
                }
            }
        }

        return result;
    }
}